### ISSUE TITLE
PM2 sporadically dies and loses processes

### ISSUE THREAD

**Issue by raine** [Unitech/pm2#2013]
It's happened a few times already. I notice that the processes are gone because metrics are not being generated and `pm2 ls` gives:

```
$ pm2 ls
[PM2] Spawning PM2 daemon
[PM2] PM2 Successfully daemonized
```

which would indicate that the PM2 daemon was dead. However there is nothing in `$HOME/.pm2/pm2.log` that would indicate why this happens. Running on OS X. I monitor load and memory usage of the machine and there is nothing interesting happening in that regard.

Any ideas?

**Comment by raine** [comment]
I may have found the reason for why this occurs. `lsof` shows a lot of entries for `node` having opened a ton of `$HOME/.pm2/rpc.sock` which could be explained by my incorrect use of the PM2 API.

**Comment by Unitech** [comment]
Make sure that you call pm2.disconnect() before exiting your app

**Comment by Unitech** [comment]
Added to PM2 documentation: https://github.com/pm2-hive/pm2-hive.github.io/commit/67ff803c2f9c0ca16410b30ea77471d541cd69ba

Would this issue be linked with #1983?

**Comment by raine** [comment]
I fixed my code by making the app connect only once and disconnect on exit, and had high hopes that it would fix the issue, however the problem just reoccurred. The pm2 main daemon had disappeared and once again, there is no meaningful output in `$HOME/.pm2/pm2.log`. Not sure how to debug this.

**Comment by fresheneesz** [comment]
Sounds like possibly the same thing as these:

https://github.com/Unitech/pm2/issues/1809
https://github.com/Unitech/pm2/issues/1320

**Comment by tuanna222** [comment]
_PM2 v2.1.5 died with these logs:_
```
2016-11-19 06:00:02: ===============================================================================
2016-11-19 06:00:02: --- PM2 global error caught ---------------------------------------------------
2016-11-19 06:00:02: Time                 : Sat Nov 19 2016 06:00:02 GMT+0700 (ICT)
2016-11-19 06:00:02: Cannot read property 'on' of undefined
2016-11-19 06:00:02: TypeError: Cannot read property 'on' of undefined
    at buildProcessTree (/usr/lib/node_modules/pm2/lib/TreeKill.js:81:14)
    at module.exports (/usr/lib/node_modules/pm2/lib/TreeKill.js:32:9)
    at Object.God.killProcess (/usr/lib/node_modules/pm2/lib/God/Methods.js:223:7)
    at Object.God.stopProcessId (/usr/lib/node_modules/pm2/lib/God/ActionMethods.js:261:9)
    at Object.God.deleteProcessId (/usr/lib/node_modules/pm2/lib/God/ActionMethods.js:313:9)
    at Worker.onListen (/usr/lib/node_modules/pm2/lib/God/Reload.js:134:18)
    at Worker.g (events.js:273:16)
    at emitOne (events.js:90:13)
    at Worker.emit (events.js:182:7)
    at listening (cluster.js:504:12)
2016-11-19 06:00:02: ===============================================================================
```

**Comment by vmarchaud** [comment]
@tuanna222 Platform ?

**Comment by tuanna222** [comment]
@vmarchaud 
Here are my info"s platform:
```
Linux app-server 2.6.32-573.el6.x86_64 #1 SMP Thu Jul 23 15:44:03 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
LSB Version:	:base-4.0-amd64:base-4.0-noarch:core-4.0-amd64:core-4.0-noarch:graphics-4.0-amd64:graphics-4.0-noarch:printing-4.0-amd64:printing-4.0-noarch
Distributor ID:	CentOS
Description:	CentOS release 6.8 (Final)
Release:	6.8
Codename:	Final
```

**Comment by Unitech** [comment]
We will do a small patch to avoid this but please note that you are using Node.js 5.x (Node v5.12.0) that is NOT a stable version. You need to upgrade your Node.js version to at least 6.x. It will certainly fix this issue.

To reproduce this issue it's like if we were using this snippet:

```javascript
const spawn = require('child_process').spawn;
const ls = spawn('ls', ['-lh', '/usr']);

ls.stdout.on('data', (data) => {
  console.log(`stdout: ${data}`);
});

ls.stderr.on('data', (data) => {
  console.log(`stderr: ${data}`);
});

ls.on('close', (code) => {
  console.log(`child process exited with code ${code}`);
});
```

And that **ls** would be null.

**Comment by tuanna222** [comment]
@vmarchaud thank to you. I updated Node.js version from 5.12.0 to 7.0.0 and it worked normally in cluster mode.

**Comment by Unitech** [comment]
Could you please have try of development branch:

```
$ npm install Unitech/pm2#development -g
$ pm2 update
```

It will work now

**Comment by tuanna222** [comment]
@Unitech 
It worked with development branch.
Thank you so much!

**Comment by ordtrogen** [comment]
Hello and thanks for looking into it. Just wanted to mention I ran into this problem yesterday and in a minute I will try the development branch. Before I do, here is a reference to my current setup (Arch Linux on an old Intel laptop)

![pm2](https://cloud.githubusercontent.com/assets/15184875/20552870/70f469d4-b14f-11e6-8327-0a6945e098a5.png)

**Comment by Unitech** [comment]
@tuanna222 great!
@ordtrogen could you please show me the content of ~/.pm2/pm2.log?

**Comment by ordtrogen** [comment]
Oh, I hadn't checked that log ...  I'm getting a lot of

2016-11-22 14:52:11: Error: ENOSPC: no space left on device, write

Which is what happened when I tried to upgrade to the development branch. I guess I should look into that first cause I don't understand how the disk can be full ...

**Comment by ordtrogen** [comment]
Oh, embarrassement ... After freeing some space, it's working again. I think I'll postpone the upgrade unless you recommend otherwise
